ZEP000-wf-stream-processing.md

```diff
@@ -0,0 +1,1421 @@
+* the `ServiceTask:Terminate_Element` (A) command transitions the service task to terminated
+* the `ServiceTask:Terminate_Element` (B) command transitions the service task to terminated, and transitions the process to terminated (because it is the last active element instance)
+
+On replay:
+* the `Process:Element_Terminating` event transitions the process to terminating 
+* the `ServiceTask:Element_Terminating` (A) transitions the service task to terminating
+* the `ServiceTask:Element_Terminated` (A) transitions the service task to terminated
+* the `ServiceTask:Element_Terminating` (B) transitions the service task to terminating
+* the `ServiceTask:Element_Terminated` (B) transitions the service task to terminated
+* the `Process:Element_Terminated` transitions the process to terminated
+
+Differences to current processing:
+* cancel the workflow instance by writing a `Process:Terminate_Element` command directly
+  * currently, a `Cancel` command is written that produces a `Process:Element_Terminating` event 
+  * omit the `Cancel` command because it doesn't have any additional business logic and don't fit to the life cycle (i.e. no `Canceled` event)
+
+Handling of concurrent commands:
+
+* if the command `Process:Terminate_Element` is written concurrently to another command of the same workflow instance (e.g. `Complete_Element` of a service task) then the other command (or a follow-up command of it) is rejected, and the workflow instance terminates eventually.
+
+* if the `Process:Terminate_Element` command is written before the `Job:Complete` command then the `ServiceTask:Complete_Element` command is rejected
+  * the `ServiceTask:Terminate_Element` command is written before the `ServiceTask:Complete_Element` command
+
+![cancel-workflow-instance](ZEP-wf-stream-processing/event-sourcing-cancel-3.png)
+
+* if the `ServiceTask:Complete_Element` command is written before the `Process:Terminate_Element` command then the `Activate_Element` of the next element is rejected
+  * on processing the `Process:Terminate_Element` command, the process transitions to terminated directly because no element instance is active (element B is not activated yet - it is just a command)
+  * reject the command `Activate_Element` of the next element because the process is already terminated
+
+![cancel-workflow-instance](ZEP-wf-stream-processing/event-sourcing-cancel-2.png)
+
+* if the `Process:Terminate_Element` command is written before the `ServiceTask:Complete_Element` command then the `ServiceTask:Complete_Element` is rejected
+  * on processing the `Process:Terminate_Element` command, the process transitions to terminating
+  * the `ServiceTask:Complete_Element` is rejected because the process (i.e. its flow scope) is on terminating
+  
+![cancel-workflow-instance](ZEP-wf-stream-processing/event-sourcing-cancel-1.png)
+
+### Incident
+
+Notes:
```

Comment:
> it feels like this atomic step is split up into a pre incident effects and post incident resolution effects

It is. Pre-incident we make sure the workflow element exists in the state (it does not exist prior to processing the `Activate_Element` command). But it needs to exist, because often incidents can only be resolved by creating some variables for that element's scope. We make sure the element exists (also on replay) when the incident is raised by writing the `Element_Activating` event, which contains all the data to store that element in state. Once the incident's cause is removed (perhaps by creating a variable or so), the incident can be resolved using `Incident:Resolve`, again processing the same command. Which may lead to another incident or to `Element_Activated`. 

>Or does raising the incident automatically roll back any of these changes?

I don't really see why we'd need some rollback. Currently, there is also no rollback on incidents (@saig0 please correct me if I'm wrong). Incidents are just treated as if the command was processed successfully. This allows the streamprocessor to continue processing other records for that partition. The state changes for that specific workflow have been partially made, but it doesn't really matter. No follow-up commands are written because the incident was raised before, so that workflow's processing does halt until the incident is resolved. On incident resolution, the workflow execution is continued again by processing the record that produced the incident again.

Reply:
Yes. But not exactly. Semantically, it is kind of a rollback. If an incident is created then no other state change is applied (e.g. creating variables, opening subscriptions). This is important because other commands can be processed before the incident is resolved.